src/ngraph/runtime/cpu/op/quantized_avg_pool.cpp

```diff
@@ -0,0 +1,232 @@
+        output_item_shape.push_back(ceil_div(input_item_virtual_shape[i] - m_window_shape[i] + 1,
+                                             m_window_movement_strides[i]));
+    }
+
+    //
+    // Make sure we're not going to have to compute average over an empty set of tensor elements.
+    // That will happen if the sliding window ever resides entirely over the padding area AND
+    // we're planning to disregard padding when computing the window's average.
+    //
+    if (!m_include_padding_in_avg_computation)
+    {
+        for (size_t i = 0; i < spatial_dimension_count; i++)
+        {
+            const size_t dim_virtual_size = input_item_virtual_shape[i];
+            const size_t dim_window_size = m_window_shape[i];
+            const size_t dim_stride = m_window_movement_strides[i];
+            const size_t dim_padding_below = m_padding_below[i];
+            const size_t dim_padding_above = m_padding_above[i];
+
+            // Checking the lower edge of each dimension is easy, because there's no mystery
+            // regarding the window's lower-edge placement...
+            NODE_VALIDATION_ASSERT(this,
+                                   dim_padding_below == 0 || dim_window_size > dim_padding_below)
+                << "Window will sometimes reside entirely within the below-padding region, but"
+                << " include_padding_in_avg_computation was not set (padding below: "
+                << m_padding_below << ", window shape: " << m_window_shape << ").";
+
+            // Now check the upper-bound...
+            {
+                const size_t dim_num_strides = (dim_virtual_size - dim_window_size) / dim_stride;
+                const size_t dim_window_max_lower_offset = dim_num_strides * dim_stride;
+                const size_t dim_padding_above_start_offset = dim_virtual_size - dim_padding_above;
+
+                NODE_VALIDATION_ASSERT(this,
+                                       dim_padding_above == 0 ||
+                                           dim_window_max_lower_offset <
+                                               dim_padding_above_start_offset)
+                    << "Window will sometimes reside entirely within the above-padding region, but"
+                    << " include_padding_in_avg_computation was not set (padding above: "
+                    << m_padding_above << ", window shape: " << m_window_shape << ").";
+            }
+        }
+    }
+
+    //
+    // Construct result shape: NCDo.
+    //
+    Shape result_shape(1 + 1 + spatial_dimension_count);
+    result_shape[0] = batch_size;
+    result_shape[1] = channel_count;
+    copy(output_item_shape.begin(), output_item_shape.end(), result_shape.begin() + 2);
+
+    set_output_size(3);
+    set_output_type(0, get_input_element_type(0), result_shape);
+    //TODO(nbpatel): Change to Shape{} once the mkldnn version is updated
+    set_output_type(1, element::f32, Shape{1});
+    set_output_type(2, element::f32, Shape{1});
```

Comment:
seems like a lot of code duplication with the core op. Can we refactor and reuse from there? Perhaps @aprocter has some thoughts.

Reply:
Heartily agree we need to refactor (probably moving duplicated logic into utility functions but maybe into subclasses). I don’t think it’s necessary for this PR, but should be done Very Soon.

Reply:
(Ideally the scope of that refactor would cover the core ops as well, which themselves have an embarrassing amount of duplication.)

Reply:
Most of the Cpu ops also have a lot of duplication from the core ops. So i guess it is a different task in itself